### PR TITLE
GNOME runtime 45

### LINKED ISSUE #29: Update runtime to 45
Consider updating the runtime version used by this Flatpak to 45.

### PR DESCRIPTION
**Pull request by AsciiWolf**
Closes #29.

### CODE REVIEW

**Comment by AsciiWolf** [comment]
Build failed on:

```
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp: In function ‘std::string MFX::printCodecId(mfxU32)’:
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:3: error: ‘uint8_t’ was not declared in this scope
   60 |   uint8_t* data = reinterpret_cast<uint8_t*>(&id);
      |   ^~~~~~~
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:29:1: note: ‘uint8_t’ is defined in header ‘<cstdint>’; did you forget to ‘#include <cstdint>’?
   28 | #include "mfxloader.h"
  +++ |+#include <cstdint>
   29 | 
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:12: error: ‘data’ was not declared in this scope
   60 |   uint8_t* data = reinterpret_cast<uint8_t*>(&id);
      |            ^~~~
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:36: error: ‘uint8_t’ does not name a type
   60 |   uint8_t* data = reinterpret_cast<uint8_t*>(&id);
      |                                    ^~~~~~~
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:36: note: ‘uint8_t’ is defined in header ‘<cstdint>’; did you forget to ‘#include <cstdint>’?
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:43: error: expected ‘>’ before ‘*’ token
   60 |   uint8_t* data = reinterpret_cast<uint8_t*>(&id);
      |                                           ^
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:43: error: expected ‘(’ before ‘*’ token
   60 |   uint8_t* data = reinterpret_cast<uint8_t*>(&id);
      |                                           ^
      |                                           (
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:44: error: expected primary-expression before ‘>’ token
   60 |   uint8_t* data = reinterpret_cast<uint8_t*>(&id);
      |                                            ^
/run/build/mediasdk/api/mfx_dispatch/linux/mfxparser.cpp:60:50: error: expected ‘)’ before ‘;’ token
   60 |   uint8_t* data = reinterpret_cast<uint8_t*>(&id);
      |                                                  ^
      |                                                  )
```

@jliljebl Any idea?

**Comment by hfiguiere** [comment]
it's just the usually transitive include issues, ie some headers need be explicitely included with newer GCC. The error message even tells you what to do.

**Comment by AsciiWolf** [comment]
You're right. :-) I am probably too overworked (or too dumb) lately...

**Comment by jliljebl** [comment]
I was unable to get this build locally. If anyone can get this build with GNOME runtime 45 I'll merge. December probably going to be quite low activity on my part but I'll try January again if we don't this working by then.

**Comment by AsciiWolf** [comment]
@jliljebl Maybe the `--enable-gpl` switch for ffmpeg module is the problem? See [this page](https://www.ffmpeg.org/legal.html).

We can try using the fd.o platform ffmpeg-full extension instead:

```
cleanup-commands:
  - mkdir -p ${FLATPAK_DEST}/lib/ffmpeg
add-extensions:
  org.freedesktop.Platform.ffmpeg-full:
    directory: lib/ffmpeg
    add-ld-path: .
    version: '23.08'
    autodownload: true
    autodelete: false
```

**Comment by AsciiWolf** [comment]
@jliljebl Could you test the new build? I am currently on a low-end laptop and cannot test it properly. Thanks!

**Comment by jliljebl** [comment]
@AsciiWolf I got something on trying to run after the install was successful

```
janne@bash~$ flatpak run io.github.jliljebl.Flowblade --user
bwrap: Can't mkdir /app/lib/ffmpeg: Read-only file system
error: ldconfig failed, exit status 256
```
you added:

```
cleanup-commands:
  - mkdir -p ${FLATPAK_DEST}/lib/ffmpeg
```
so that's running at the time of first launch. I thought stuff in */app/lib/\** would be immutable after install until doing uninstall.

File: io.github.jliljebl.Flowblade.yaml [hunk 1]
```diff
@@ -13,22 +13,30 @@ finish-args:
   - --env=FREI0R_PATH=/app/lib/frei0r-1
   - --env=LADSPA_PATH=/app/extensions/Plugins/ladspa:/app/lib/ladspa
   - --talk-name=org.freedesktop.Flatpak
+cleanup-commands:
+  - mkdir -p ${FLATPAK_DEST}/lib/ffmpeg
```
[the remark this answers is not in the document]
Comment (AsciiWolf) [reply io.github.jliljebl.Flowblade.yaml]: Done. Thanks!

**Comment by AsciiWolf** [comment]
It runs now and seems to work, however trying to render the project results in some tracebacks being printed on stdout:

```
Traceback (most recent call last):
  File "/app/share/flowblade/Flowblade/editorwindow.py", line 882, in <lambda>
    ('AddToQueue', None, _('Add To Batch Render Queue...'), None, None, lambda a:projectaction.add_to_render_queue()),
                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/projectaction.py", line 829, in add_to_render_queue
    _write_out_render_item(False)
  File "/app/share/flowblade/Flowblade/projectaction.py", line 837, in _write_out_render_item
    args_vals_list = render.get_args_vals_list_for_current_selections()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/render.py", line 75, in get_args_vals_list_for_current_selections
    encoding_option_index = widgets.encoding_panel.encoding_selector.get_selected_encoding_index()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/rendergui.py", line 532, in get_selected_encoding_index
    name, encoding = self.categorised_combo.get_selected()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/guicomponents.py", line 3702, in get_selected
    indices = self.model.get_path(self.widget.get_active_iter()).get_indices()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Argument 1 does not allow None as a value
Saving project...
Saving project...
Traceback (most recent call last):
  File "/app/share/flowblade/Flowblade/editorwindow.py", line 885, in <lambda>
    ('Render', None, _('Render Timeline'), None, None, lambda a:projectaction.do_rendering()),
                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/projectaction.py", line 802, in do_rendering
    success = _write_out_render_item(True)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/projectaction.py", line 837, in _write_out_render_item
    args_vals_list = render.get_args_vals_list_for_current_selections()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/render.py", line 75, in get_args_vals_list_for_current_selections
    encoding_option_index = widgets.encoding_panel.encoding_selector.get_selected_encoding_index()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/rendergui.py", line 532, in get_selected_encoding_index
    name, encoding = self.categorised_combo.get_selected()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/guicomponents.py", line 3702, in get_selected
    indices = self.model.get_path(self.widget.get_active_iter()).get_indices()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Argument 1 does not allow None as a value
Saving project...
Saving project...
FileRenderPlayer started, start frame: 66, stop frame: 67
FileRenderPlayer stopped, producer frame: 68
Saving project...
Traceback (most recent call last):
  File "/app/share/flowblade/Flowblade/editorwindow.py", line 882, in <lambda>
    ('AddToQueue', None, _('Add To Batch Render Queue...'), None, None, lambda a:projectaction.add_to_render_queue()),
                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/projectaction.py", line 829, in add_to_render_queue
    _write_out_render_item(False)
  File "/app/share/flowblade/Flowblade/projectaction.py", line 837, in _write_out_render_item
    args_vals_list = render.get_args_vals_list_for_current_selections()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/render.py", line 75, in get_args_vals_list_for_current_selections
    encoding_option_index = widgets.encoding_panel.encoding_selector.get_selected_encoding_index()
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/rendergui.py", line 532, in get_selected_encoding_index
    name, encoding = self.categorised_combo.get_selected()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/share/flowblade/Flowblade/guicomponents.py", line 3702, in get_selected
    indices = self.model.get_path(self.widget.get_active_iter()).get_indices()
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: Argument 1 does not allow None as a value
Exception in thread Thread-7:
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/app/share/flowblade/Flowblade/singletracktransition.py", line 709, in run
    self.rerender_window.render_next()
  File "/app/share/flowblade/Flowblade/singletracktransition.py", line 600, in render_next
    clip, track = self.rerender_list[self.current_item]
                  ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
IndexError: list index out of range
```

And the actual rendering process seems to do nothing. But it may be some problem related to my low-end laptop GPU.

@jliljebl Could you retest the new build? Thanks!

**Comment by jliljebl** [comment]
@AsciiWolf thanks for the update. Now application launches, and I was able to do a render. I think the: 

```
Traceback (most recent call last):
  File "/usr/lib/python3.11/threading.py", line 1038, in _bootstrap_inner
    self.run()
  File "/app/share/flowblade/Flowblade/singletracktransition.py", line 709, in run
    self.rerender_window.render_next()
  File "/app/share/flowblade/Flowblade/singletracktransition.py", line 600, in render_next
    clip, track = self.rerender_list[self.current_item]
                  ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
IndexError: list index out of range
```
is because we lost the *H.264 / mp4* encoding option with the update.

So there are some issues.

1) LADSPA filters missing 
```
MLT service ladspa.1213 not found.
MLT service ladspa.1904 not found.
MLT service ladspa.1903 not found.
...
```
2) A lot of encoding options are missing now like h264, HEVC an NVENC that was present in my system.
I think this is related to ffmpeg being updated from 5.x -> 6.0 which in my system also resulted in losing NVENC stuff.

I believe the situation is now that the flatpak update to GNOME runtime 45 and ffmpeg 6.0 works fine but Flowblade encoding parameters and maybe MLT need to be updated to not lose functionality when doing this update.

Unfortunately I have an annoying flu going on right now and will get back to bed after writing this comment. I'll post next update on on Tuesday, hopefully flu has cleared up by then.

**Comment by AsciiWolf** [comment]
No problem. I am glad I could help. :-)

Regarding Intel Media SDK (that was [discontinued](https://github.com/Intel-Media-SDK/MediaSDK/commit/7a72de33a15d6e7cdb842b12b901a003f7154f0a) this year), it looks like that Kdenlive is switching to using Intel oneVPL instead:

https://github.com/flathub/org.kde.kdenlive/pull/292/files#diff-42a5057622e628a20adb9f2d8f5123a5b0144d93f06818a8e86fa64ab81d1fd5L575

Maybe it could also be used with Flowblade?

**Comment by jliljebl** [comment]
> https://github.com/flathub/org.kde.kdenlive/pull/292/files#diff-42a5057622e628a20adb9f2d8f5123a5b0144d93f06818a8e86fa64ab81d1fd5L575

> Maybe it could also be used with Flowblade?

Thanks, this is very likely the way to go.

**Pull request closed without merging**